NoSuchMethodError with AGP 3.6.1 and marathon 0.5.1

Looks like something is broken with latest Android Gradle Plugin.

```
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':songsterr:marathonGplayDebugAndroidTest'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:205)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:203)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:184)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:114)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.NoSuchMethodError: com.android.build.gradle.tasks.PackageAndroidArtifact.getOutputDirectory()Ljava/io/File;
        at com.malinskiy.marathon.extensions.TestVariantKt.extractTestApplicationBefore3_3(TestVariant.kt:27)
        at com.malinskiy.marathon.extensions.TestVariantKt.access$extractTestApplicationBefore3_3(TestVariant.kt:1)
        at com.malinskiy.marathon.extensions.TestVariantKt$extractTestApplication$2.invoke(TestVariant.kt:17)
        at com.malinskiy.marathon.extensions.TestVariantKt$extractTestApplication$2.invoke(TestVariant.kt)
        at com.malinskiy.marathon.extensions.GradleExtensionsKt.executeGradleCompat(GradleExtensions.kt:7)
        at com.malinskiy.marathon.extensions.TestVariantKt.extractTestApplication(TestVariant.kt:12)
        at com.malinskiy.marathon.MarathonRunTask.runMarathon(MarathonRunTask.kt:41)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
        ... 30 more

```

--- a/marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/MarathonPlugin.kt
+++ b/marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/MarathonPlugin.kt
@@ -61,7 +61,13 @@ class MarathonPlugin : Plugin<Project> {
     }
 
     companion object {
-        private fun createTask(variant: TestVariant, project: Project, config: MarathonExtension, sdkDirectory: File, exceptionsReporter: ExceptionsReporter): MarathonRunTask {
+        private fun createTask(
+            variant: TestVariant,
+            project: Project,
+            config: MarathonExtension,
+            sdkDirectory: File,
+            exceptionsReporter: ExceptionsReporter
+        ): MarathonRunTask {
             checkTestVariants(variant)
 
             val marathonTask = project.tasks.create("$TASK_PREFIX${variant.name.capitalize()}", MarathonRunTask::class.java)
@@ -74,7 +80,7 @@ class MarathonPlugin : Plugin<Project> {
                 marathonTask.configure(closureOf<MarathonRunTask> {
                     group = JavaBasePlugin.VERIFICATION_GROUP
                     description = "Runs instrumentation tests on all the connected devices for '${variant.name}' " +
-                            "variation and generates a report with screenshots"
+                        "variation and generates a report with screenshots"
                     flavorName = variant.name
                     applicationVariant = variant.testedVariant
                     testVariant = variant
@@ -86,7 +92,7 @@ class MarathonPlugin : Plugin<Project> {
                         exec = {
                             dependsOn(variant.testedVariant.assembleProvider, variant.assembleProvider)
                         },
-                        fallback = {
+                        fallbacks = listOf {
                             @Suppress("DEPRECATION")
                             dependsOn(variant.testedVariant.assemble, variant.assemble)
                         }
@@ -114,8 +120,8 @@ class MarathonPlugin : Plugin<Project> {
             if (baseVariantOutput.outputs.size > 1) {
                 throw UnsupportedOperationException(
                     "The Marathon plugin does not support abi splits for app APKs, " +
-                            "but supports testing via a universal APK. "
-                            + "Add the flag \"universalApk true\" in the android.splits.abi configuration."
+                        "but supports testing via a universal APK. "
+                        + "Add the flag \"universalApk true\" in the android.splits.abi configuration."
                 )
             }
 

--- a/marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/extensions/BaseVariant.kt
+++ b/marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/extensions/BaseVariant.kt
@@ -12,7 +12,7 @@ fun BaseVariant.extractApplication(): File? =
         exec = {
             extractApplication3_3_plus(this)
         },
-        fallback = {
+        fallbacks = listOf {
             extractApplicationBefore3_3(this)
         }
     )

--- a/marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/extensions/GradleExtensions.kt
+++ b/marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/extensions/GradleExtensions.kt
@@ -1,9 +1,17 @@
 package com.malinskiy.marathon.extensions
 
-fun <T> executeGradleCompat(exec: () -> T, fallback: () -> T): T {
+fun <T> executeGradleCompat(exec: () -> T, fallbacks: List<() -> T>): T {
     return try {
         exec.invoke()
     } catch (e: NoSuchMethodError) {
-        return fallback.invoke()
+        for (fallback in fallbacks) {
+            try {
+                return fallback.invoke()
+            } catch (e: Exception) {
+                //Try next fallback
+            }
+        }
+
+        throw RuntimeException("Unable to execute gradle compat: all fallbacks failed")
     }
 }

--- a/marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/extensions/TestVariant.kt
+++ b/marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/extensions/TestVariant.kt
@@ -9,14 +9,39 @@ import org.apache.tools.ant.taskdefs.Zip
 import org.gradle.api.GradleException
 import java.io.File
 
-fun TestVariant.extractTestApplication() = com.malinskiy.marathon.extensions.executeGradleCompat(
+fun TestVariant.extractTestApplication() = executeGradleCompat(
     exec = {
-        extractTestApplication3_3_plus(this)
+        extractTestApplication3_6_plus(this)
     },
-    fallback = {
-        extractTestApplicationBefore3_3(this)
-    }
+    fallbacks = listOf(
+        {
+            extractTestApplication3_3_to_3_5(this)
+        },
+        {
+            extractTestApplicationBefore3_3(this)
+        }
+    )
 )
+
+fun extractTestApplication3_6_plus(variant: TestVariant): File {
+    val output = variant.outputs.first()
+
+    return File(
+        when (output) {
+            is ApkVariantOutput -> {
+                val packageTask =
+                    variant.packageApplicationProvider.orNull ?: throw IllegalArgumentException("Can't find package application provider")
+                File(packageTask.outputDirectory.asFile.get(), output.outputFileName).path
+            }
+            is LibraryVariantOutput -> {
+                output.outputFile.path
+            }
+            else -> {
+                throw RuntimeException("Can't find instrumentationApk")
+            }
+        }
+    )
+}
 
 private fun extractTestApplicationBefore3_3(variant: TestVariant): File {
     val output = variant.outputs.first()
@@ -24,6 +49,7 @@ private fun extractTestApplicationBefore3_3(variant: TestVariant): File {
     return File(
         when (output) {
             is ApkVariantOutput -> {
+                variant.packageApplicationProvider
                 File(variant.packageApplication.outputDirectory.asFile.get(), output.outputFileName).path
             }
             is LibraryVariantOutput -> {
@@ -36,7 +62,7 @@ private fun extractTestApplicationBefore3_3(variant: TestVariant): File {
     )
 }
 
-private fun extractTestApplication3_3_plus(output: TestVariant): File {
+private fun extractTestApplication3_3_to_3_5(output: TestVariant): File {
     val testPackageAndroidArtifact = when (output) {
         is TestVariant -> {
             output.packageApplicationProvider

--- a/sample/android-app/build.gradle.kts
+++ b/sample/android-app/build.gradle.kts
@@ -7,6 +7,15 @@ buildscript {
     dependencies {
         classpath(BuildPlugins.kotlinPlugin)
         classpath(BuildPlugins.androidGradle)
+        /**
+         * Starting with kotlin plugin 1.3.41 coroutines dependency is not propagated to the classpath of gradle plugin
+         *
+         * e.g.
+         * Caused by: java.lang.NoSuchMethodError: kotlinx.coroutines.channels.ChannelIterator.next()Ljava/lang/Object;
+         *
+         * Hence we need to explicitly add coroutines to our classpath
+         */
+        classpath(Libraries.kotlinCoroutines)
     }
 }
 

--- a/sample/android-app/buildSrc/src/main/kotlin/Versions.kt
+++ b/sample/android-app/buildSrc/src/main/kotlin/Versions.kt
@@ -1,5 +1,6 @@
 object Versions {
     val kotlin = "1.3.61"
+    val coroutines = "1.3.3"
 
     val androidGradleVersion = "3.6.2"
 
@@ -22,6 +23,7 @@ object BuildPlugins {
 object Libraries {
     val kotlinStdLib = "org.jetbrains.kotlin:kotlin-stdlib-jdk8:${Versions.kotlin}"
     val kotlinReflect = "org.jetbrains.kotlin:kotlin-reflect:${Versions.kotlin}"
+    val kotlinCoroutines = "org.jetbrains.kotlinx:kotlinx-coroutines-core:${Versions.coroutines}"
     val appCompat = "com.android.support:appcompat-v7:${Versions.appCompat}"
     val constraintLayout = "com.android.support.constraint:constraint-layout:${Versions.constraintLayout}"
 }

--- a/sample/android-library/build.gradle.kts
+++ b/sample/android-library/build.gradle.kts
@@ -7,6 +7,15 @@ buildscript {
     dependencies {
         classpath(BuildPlugins.kotlinPlugin)
         classpath(BuildPlugins.androidGradle)
+        /**
+         * Starting with kotlin plugin 1.3.41 coroutines dependency is not propagated to the classpath of gradle plugin
+         *
+         * e.g.
+         * Caused by: java.lang.NoSuchMethodError: kotlinx.coroutines.channels.ChannelIterator.next()Ljava/lang/Object;
+         *
+         * Hence we need to explicitly add coroutines to our classpath
+         */
+        classpath(Libraries.kotlinCoroutines)
     }
 }
 

--- a/sample/android-library/buildSrc/src/main/kotlin/Versions.kt
+++ b/sample/android-library/buildSrc/src/main/kotlin/Versions.kt
@@ -1,5 +1,6 @@
 object Versions {
     val kotlin = "1.3.61"
+    val coroutines = "1.3.3"
 
     val androidGradleVersion = "3.6.2"
 
@@ -20,6 +21,7 @@ object BuildPlugins {
 object Libraries {
     val kotlinStdLib = "org.jetbrains.kotlin:kotlin-stdlib-jdk8:${Versions.kotlin}"
     val kotlinReflect = "org.jetbrains.kotlin:kotlin-reflect:${Versions.kotlin}"
+    val kotlinCoroutines = "org.jetbrains.kotlinx:kotlinx-coroutines-core:${Versions.coroutines}"
     val appCompat = "com.android.support:appcompat-v7:${Versions.appCompat}"
     val constraintLayout = "com.android.support.constraint:constraint-layout:${Versions.constraintLayout}"
 }

